vizro-ai/src/vizro_ai/dashboard/_response_models/page.py

```diff
@@ -0,0 +1,222 @@
+        component_log.close()
+        return components
+
+    def _get_layout(self, model, all_df_metadata):
+        if self._layout is None:
+            self._layout = self._build_layout(model, all_df_metadata)
+        return self._layout
+
+    def _build_layout(self, model, all_df_metadata):
+        if self.layout_plan is None:
+            return None
+        return self.layout_plan.create(
+            component_ids=self._get_component_ids(model=model, all_df_metadata=all_df_metadata),
+        )
+
+    def _get_controls(self, model, all_df_metadata):
+        if self._controls is None:
+            self._controls = self._build_controls(model=model, all_df_metadata=all_df_metadata)
+        return self._controls
+
+    def _controllable_components(self, model, all_df_metadata):
+        return [
+            comp.id
+            for comp in self._get_components(model=model, all_df_metadata=all_df_metadata)
+            if isinstance(comp, (vm.Graph, vm.AgGrid))
+        ]
+
+    def _get_component_ids(self, model, all_df_metadata):
+        return [comp.id for comp in self._get_components(model=model, all_df_metadata=all_df_metadata)]
+
+    def _build_controls(self, model, all_df_metadata):
+        controls = []
+        with tqdm(
+            total=len(self.controls_plan),
+            desc=f"Currently Building ... [Page] <{self.title}> controls",
+            leave=False,
+        ) as pbar:
+            for control_plan in self.controls_plan:
+                pbar.update(1)
+                control = control_plan.create(
+                    model=model,
+                    controllable_components=self._controllable_components(model=model, all_df_metadata=all_df_metadata),
+                    all_df_metadata=all_df_metadata,
+                )
+                if control:
+                    controls.append(control)
+
+        return controls
+
+    def create(self, model, all_df_metadata) -> Union[vm.Page, None]:
+        """Create the page."""
+        page_desc = f"Building page: {self.title}"
+        logger.info(page_desc)
+        pbar = tqdm(total=5, desc=page_desc)
+
+        title = _execute_step(pbar, page_desc + " --> add title", self.title)
+        components = _execute_step(
+            pbar, page_desc + " --> add components", self._get_components(model=model, all_df_metadata=all_df_metadata)
+        )
+        controls = _execute_step(
+            pbar, page_desc + " --> add controls", self._get_controls(model=model, all_df_metadata=all_df_metadata)
+        )
+        layout = _execute_step(
+            pbar, page_desc + " --> add layout", self._get_layout(model=model, all_df_metadata=all_df_metadata)
+        )
+
+        try:
+            page = vm.Page(title=title, components=components, controls=controls, layout=layout)
+        except Exception as e:
+            if any("Number of page and grid components need to be the same" in error["msg"] for error in e.errors()):
```

Comment:
Can we not catch this earlier, when the layout is created?

Reply:
Good question. I created the ticket for next sprint to revisit this. Ideally we should return no layout if failed.